airflow/models/base.py

```diff
@@ -34,9 +41,123 @@
 ID_LEN = 250
 
 
-# used for typing
 class Operator:
-    """Class just used for Typing"""
+    """Common interface for operators, including unmapped and mapped."""
+
+    log: logging.Logger
+
+    upstream_task_ids: Set[str]
+    downstream_task_ids: Set[str]
+    weight_rule: str
+    priority_weight: int
+
+    # For derived classes to define which fields will get jinjaified.
+    template_fields: Collection[str]
+    # Defines which files extensions to look for in the templated fields.
+    template_ext: Collection[str]
+
+    def get_dag(self) -> "Optional[DAG]":
+        raise NotImplementedError()
+
+    @property
+    def dag_id(self) -> str:
+        """Returns dag id if it has one or an adhoc + owner"""
+        dag = self.get_dag()
```

Comment:
I think this needs to use `has_dag`, otherwise for a BaseOperator without a dag this would throw an exception (and never hit the "adhoc_" branch.

Reply:
Oh, no, that's what `self.dag` would do, this is fine.